src/shims/native_lib/mod.rs

```diff
@@ -169,6 +211,52 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
         }
         None
     }
+
+    /// Applies the `events` to Miri's internal state. The event vector must be
+    /// ordered sequentially by when the accesses happened, and the sizes are
+    /// assumed to be exact.
+    fn tracing_apply_accesses(&mut self, events: MemEvents) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        for evt in events.acc_events {
+            let evt_rg = evt.get_range();
+            // We're assuming an access only touches 1 allocation.
+            let alloc_id = this
+                .alloc_id_from_addr(evt_rg.addr.to_u64(), evt_rg.size.try_into().unwrap(), true)
+                .expect("Foreign code did an out-of-bounds access!");
+
+            // Get the (size, len) pair for the current allocation.
+            let (alloc_addr, alloc_len) = {
+                let alloc = this.get_alloc_raw(alloc_id)?;
+                (alloc.get_bytes_unchecked_raw().addr(), alloc.len())
+            };
+
+            let unshifted_overlap = std::cmp::max(evt_rg.addr, alloc_addr)
+                ..std::cmp::min(
+                    evt_rg.addr.strict_add(evt_rg.size),
+                    alloc_addr.strict_add(alloc_len),
+                );
```

Comment:
Given that we assert 1 access = 1 allocation, should we skip this entirely and just error if the access is out of bounds?